src/shims/unix/linux/epoll.rs

```diff
@@ -413,34 +417,40 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         if ready_list.is_empty() {
             if timeout == 0 {
                 // Non-blocking with no notification returned.
-                this.write_scalar(Scalar::from_i32(0), &des)?;
+                this.write_scalar(Scalar::from_i32(0), &dest)?;
                 return Ok(());
             } else {
                 // Blocking
-                let duration = Duration::from_secs(timeout.into());
+                let duration = Duration::from_millis(timeout.into());
+                drop(ready_list);
                 this.block_thread(
                     BlockReason::Epoll,
                     Some((TimeoutClock::Monotonic, TimeoutAnchor::Relative, duration)),
                     callback!(
                         @capture<'tcx> {
                             epfd_value: i32,
-                            des: MPlaceTy<'tcx>,
+                            weak_epfd: WeakFileDescriptionRef,
+                            dest: MPlaceTy<'tcx>,
                             event: MPlaceTy<'tcx>,
                         }
                         @unblock = |this| {
-                            this.blocking_epoll_callback(epfd_value, &des, &event)?;
+                            this.blocking_epoll_callback(epfd_value, weak_epfd, &dest, &event)?;
                             Ok(())
                         }
                         @timeout = |this| {
-                            this.write_scalar(Scalar::from_i32(0), &des)?;
+                            // No notification after blocking timeout.
+                            this.write_scalar(Scalar::from_i32(0), &dest)?;
                             Ok(())
                         }
                     ),
                 );
             }
         } else {
-            // Non-blocking with notification returned.
-            this.blocking_epoll_callback(epfd_value, &des, &event)?;
+            // Immediately return epoll notification without blocking.
+
+            // Attempt to prevent epfd.borrow_mut in blocking_epoll_callback causing ICE.
+            drop(ready_list);
```

Comment:
You need to drop `epfd`, which may be hard because of other things that need to be dropped first which are shadowed (e.g. `binding`). Maybe just move everything from the first `borrow_mut` into a nested block and only declare `weak_epfd` before the block

Reply:
Good idea, it works!